src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,43 +1,52 @@
 package com.es.phoneshop.model.product;
 
+import com.es.phoneshop.model.exception.ProductNotFoundException;
+
 import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
-    List<Product> products;
+    private List<Product> products;
 
-    public ArrayListProductDao() {
+    ArrayListProductDao() {
         products = new ArrayList<>();
     }
 
-
     @Override
-    public Product getProduct(Long id) {
+    public synchronized Product getProduct(Long id) {
         return products.stream().
                 filter(product -> product.getId().equals(id)).
-                findAny().orElseThrow( () -> new RuntimeException( "Product is not found"));
+                findAny().orElseThrow( () -> new ProductNotFoundException( "Product is not found"));
```

Comment:
Here and in other places - formatting:

https://marcin-chwedczuk.github.io/java-streams-best-practices

Compare readability:

return products.stream().
                filter(product -> product.getId().equals(id)).
                findAny().orElseThrow(...)

VS:

return products.stream()
                .filter(product -> product.getId().equals(id))
                .findAny()
                .orElseThrow(...)